### PR TITLE
Properly serialize enums to JSON in Swift implementation

### PR DESCRIPTION
**Pull request by efirestone**
Match the JSON spec for serialization of enums: https://developers.google.com/protocol-buffers/docs/proto3#json

### CODE REVIEW

File: wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public init(from decoder: Decoder) throws {
+        // We support decoding from either the string value or the field index.
+        let container = try decoder.singleValueContainer()
+        do {
+            let string = try container.decode(String.self)
+            guard let value = T.allCases.first(where: { "\($0)" == string }) else {
+                throw ProtoDecoder.Error.unknownEnumString(type: T.self, string: string)
+            }
+            self.wrappedValue = value
+        } catch {
```
Comment (JakeWharton) [inline wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift:40]: Won't this catch the ProtoDecoder.Error that's thrown? Is this meant to catch only the decoding as string step?

File: wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            let fieldNumber = try container.decode(UInt32.self)
+            guard let value = T.init(rawValue: fieldNumber) else {
+                throw ProtoDecoder.Error.unknownEnumCase(type: T.self, fieldNumber: fieldNumber)
+            }
+            self.wrappedValue = value
+        }
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        try "\(wrappedValue)".encode(to: encoder)
```
Comment (JakeWharton) [inline wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift:51]: Is this efficient? I remember in Kotlin that things like `"$this"` are short but wind up creating more machinery than the slightly-more verbose `this.toString()`.

File: wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift (hunk 1 above)
Comment (JakeWharton) [reply wire-library/wire-runtime-swift/src/main/swift/JSONEnum.swift:40]: Can we add test cases for these enum string mismatch and enum uint mismatch cases to validate the right error is propagated?

**Review by JakeWharton — APPROVED** [review]

**Comment by oldergod** [comment]
@efirestone That'd be sweet to finish this one

**Pull request closed without merging**